### PR TITLE
NetPlay: Fix hosting being stuck if player leaves during pending start

### PR DESCRIPTION
**Pull request by Techjar**
The logic didn't account for the case where a player leaves, so the host would be left in a dangling state where the UI is disabled but the game won't start, requiring a full restart of Dolphin to fix.

I've also added a cancel button to the chunked transfer dialog, since that's possible now.

### CODE REVIEW

File: Source/Core/Core/NetPlayClient.cpp [hunk 1]
```diff
@@ -365,14 +368,17 @@ unsigned int NetPlayClient::OnData(sf::Packet& packet)
...
 
-    OnData(m_chunked_data_receive_queue[cid]);
-    m_chunked_data_receive_queue.erase(m_chunked_data_receive_queue.find(cid));
-    m_dialog->HideChunkedProgressDialog();
-
-    sf::Packet complete_packet;
-    complete_packet << static_cast<MessageId>(NP_MSG_CHUNKED_DATA_COMPLETE);
-    complete_packet << cid;
-    Send(complete_packet, CHUNKED_DATA_CHANNEL);
+    if (m_chunked_data_receive_queue.find(cid) != m_chunked_data_receive_queue.end())
```
Comment (jordan-woyak) [inline Source/Core/Core/NetPlayClient.cpp]: m_chunked_data_receive_queue.count(cid)

File: Source/Core/Core/NetPlayClient.cpp [hunk 2]
```diff
@@ -365,14 +368,17 @@ unsigned int NetPlayClient::OnData(sf::Packet& packet)
...
+    {
+      OnData(m_chunked_data_receive_queue[cid]);
+      m_chunked_data_receive_queue.erase(m_chunked_data_receive_queue.find(cid));
```
Comment (jordan-woyak) [inline Source/Core/Core/NetPlayClient.cpp]: m_chunked_data_receive_queue.erase(cid);

File: Source/Core/Core/NetPlayClient.cpp [hunk 3]
```diff
@@ -381,21 +387,37 @@ unsigned int NetPlayClient::OnData(sf::Packet& packet)
     u32 cid;
     packet >> cid;
 
-    while (!packet.endOfPacket())
+    if (m_chunked_data_receive_queue.find(cid) != m_chunked_data_receive_queue.end())
```
Comment (jordan-woyak) [inline Source/Core/Core/NetPlayClient.cpp]: m_chunked_data_receive_queue.count(cid)

File: Source/Core/Core/NetPlayClient.cpp [hunk 4]
```diff
@@ -381,21 +387,37 @@ unsigned int NetPlayClient::OnData(sf::Packet& packet)
...
+  {
+    u32 cid;
+    packet >> cid;
 
-    sf::Packet progress_packet;
-    progress_packet << static_cast<MessageId>(NP_MSG_CHUNKED_DATA_PROGRESS);
-    progress_packet << cid;
-    progress_packet << sf::Uint64{m_chunked_data_receive_queue[cid].getDataSize()};
-    Send(progress_packet, CHUNKED_DATA_CHANNEL);
+    if (m_chunked_data_receive_queue.find(cid) != m_chunked_data_receive_queue.end())
```
Comment (jordan-woyak) [inline Source/Core/Core/NetPlayClient.cpp]: m_chunked_data_receive_queue.count(cid)

File: Source/Core/Core/NetPlayClient.cpp [hunk 5]
```diff
@@ -381,21 +387,37 @@ unsigned int NetPlayClient::OnData(sf::Packet& packet)
...
+    {
+      m_chunked_data_receive_queue.erase(m_chunked_data_receive_queue.find(cid));
```
Comment (jordan-woyak) [inline Source/Core/Core/NetPlayClient.cpp]: m_chunked_data_receive_queue.erase(cid);

File: Source/Core/Core/NetPlayClient.cpp (hunk 2 above)
Comment (Techjar) [reply Source/Core/Core/NetPlayClient.cpp]: `erase` takes an iterator position, not an element.

Comment (jordan-woyak) [reply Source/Core/Core/NetPlayClient.cpp]: There is an overload for the key type.

File: Source/Core/Core/NetPlayClient.cpp (hunk 1 above)
Comment (Techjar) [reply Source/Core/Core/NetPlayClient.cpp]: That is functionally equivalent, and `find` is faster when multiple elements of the same value exist, so I use this method as standard practice.

File: Source/Core/Core/NetPlayClient.cpp (hunk 2 above)
Comment (Techjar) [reply Source/Core/Core/NetPlayClient.cpp]: The documentation doesn't seem to think so.

**Pull request merged**